sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java

```diff
@@ -262,6 +265,8 @@ public static void main(
                     }
                   });
 
+      MetricsEnvironment.setProcessWideContainer(new MetricsContainerImpl());
```

Comment:
Remind me why MetricsEnvironment needs a setter rather than it just implicitly always creating one.

Reply:
The main reason was to allow clearing/cleaning the state between tests.

Explicitly creating it forces you to create a clean state on startup.
We could have had one here by default as well. I don't feel too strongly here.
Though I would consider it out of scope of this PR

Reply:
My preference is to have the tests be non-idiomatic (e.g. you could have a reset-process-container-for-test method) than have the "production" code be non-idiomatic to accommodate tests. Having a setter doesn't prevent state from leaking from test to test (though I guess it forces you to set it in at least one test).

Reply:
Okay, I looked into this a bit more. And I don't think that I can set the MetricContainerInmpl as the default container in MetricsEnvironment without adding a bad dependency, which I think may introduce a cycle.

As intellij seems to want to add the dependency when I try to instantiate the MetricsContainerImpl in MetricsEnvironment.
I dont think we can add a dep in this direction. (Today, MetricsEnvironment only depends on the interface MetricsContainer).

sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsEnvironment.java
runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java

That is, we don't make the sdk/java/core depend on the runners/core-java.
The way it works today is because we allow sdks/java/harness to depend on runners/core-java.

which is why we can instantiate it in 
sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java

Some larger refactor may make this possible, but I feel its unclear what would be the best choice here and would be too big a refactor for this PR

Reply:
OK, I agree a refactor seems out of scope for this PR.